src/app/clusters/microwave-oven-control-server/microwave-oven-control-server.cpp

```diff
@@ -180,19 +322,51 @@ void Instance::HandleSetCookingParameters(HandlerContext & ctx, const Commands::
                  ChipLogError(Zcl, "Microwave Oven Control: Failed to set cookMode, cookMode is not supported"));
 
     reqCookTime = cookTime.ValueOr(MicrowaveOvenControl::kDefaultCookTime);
-    VerifyOrExit(IsCookTimeInRange(reqCookTime), status = Status::InvalidCommand;
+    VerifyOrExit(IsCookTimeInRange(reqCookTime, mDelegate->GetMaxCookTime()), status = Status::ConstraintError;
                  ChipLogError(Zcl, "Microwave Oven Control: Failed to set cookTime, cookTime value is out of range"));
 
-    reqPowerSetting = powerSetting.ValueOr(mDelegate->GetMaxPower());
-    VerifyOrExit(IsPowerSettingInRange(reqPowerSetting, mDelegate->GetMinPower(), mDelegate->GetMaxPower()),
-                 status = Status::InvalidCommand;
-                 ChipLogError(Zcl, "Microwave Oven Control: Failed to set cookPower, cookPower value is out of range"));
+    if (HasFeature(MicrowaveOvenControl::Feature::kPowerAsNumber)) // power as number
+    {
+        VerifyOrExit(!wattSettingIndex.HasValue(), status = Status::InvalidCommand; ChipLogError(
+            Zcl, "Microwave Oven Control: Failed to set cooking parameters, should have no value for wattSettingIndex"));
+
+        VerifyOrExit(
+            cookMode.HasValue() || cookTime.HasValue() || powerSetting.HasValue(), status = Status::InvalidCommand;
+            ChipLogError(Zcl, "Microwave Oven Control: Failed to set cooking parameters, all command fields are missing "));
+
+        if (HasFeature(MicrowaveOvenControl::Feature::kPowerNumberLimits))
+        {
+            maxPower = mDelegate->GetMaxPower();
+            minPower = mDelegate->GetMinPower();
+        }
+        reqPowerSetting = powerSetting.ValueOr(maxPower);
+        VerifyOrExit(IsPowerSettingInRange(reqPowerSetting, minPower, maxPower), status = Status::ConstraintError;
+                     ChipLogError(Zcl, "Microwave Oven Control: Failed to set cookPower, cookPower value is out of range"));
+
+        status = mDelegate->HandleSetCookingParametersCallback(reqCookMode, reqCookTime, reqStartAfterSetting,
+                                                               MakeOptional(reqPowerSetting), NullOptional);
+    }
+    else // power in watts
+    {
+        VerifyOrExit(!powerSetting.HasValue(), status = Status::InvalidCommand; ChipLogError(
+            Zcl, "Microwave Oven Control: Failed to set cooking parameters, should have no value for powerSetting "));
+
+        VerifyOrExit(
+            cookMode.HasValue() || cookTime.HasValue() || wattSettingIndex.HasValue(), status = Status::InvalidCommand;
+            ChipLogError(Zcl, "Microwave Oven Control: Failed to set cooking parameters, all command fields are missing "));
 
-    status = mDelegate->HandleSetCookingParametersCallback(reqCookMode, reqCookTime, reqPowerSetting);
+        reqWattSettingIndex = wattSettingIndex.ValueOr(mSupportedPowerLevels - 1);
```

Comment:
should we assert that mSupportedPowerLevels > 0  to allow for the substraction on unsigned values?